rpcs3/Emu/Audio/XAudio2/xaudio2_enumerator.cpp

```diff
@@ -0,0 +1,145 @@
+#ifndef _WIN32
+#error "XAudio2 can only be built on Windows."
+#endif
+
+#include "Emu/Audio/XAudio2/xaudio2_enumerator.h"
+#include "util/logs.hpp"
+#include "Utilities/StrUtil.h"
+
+#include <wrl/client.h>
+#include <Windows.h>
+#include <system_error>
+#include <mmdeviceapi.h>
+#include <Functiondiscoverykeys_devpkey.h>
+
+LOG_CHANNEL(xaudio_dev_enum);
+
+xaudio2_enumerator::xaudio2_enumerator() : audio_device_enumerator()
+{
+}
+
+xaudio2_enumerator::~xaudio2_enumerator()
+{
+}
+
+std::vector<audio_device_enumerator::audio_device> xaudio2_enumerator::get_output_devices()
+{
+	Microsoft::WRL::ComPtr<IMMDeviceEnumerator> devEnum{};
+	if (HRESULT hr = CoCreateInstance(__uuidof(MMDeviceEnumerator), nullptr, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(devEnum.GetAddressOf())); FAILED(hr))
+	{
+		xaudio_dev_enum.error("CoCreateInstance() failed: %s (0x%08x)", std::system_category().message(hr), static_cast<u32>(hr));
+		return {};
+	}
+
+	Microsoft::WRL::ComPtr<IMMDeviceCollection> devices{};
+	if (HRESULT hr = devEnum->EnumAudioEndpoints(eRender, DEVICE_STATE_ACTIVE | DEVICE_STATE_DISABLED | DEVICE_STATE_UNPLUGGED, &devices); FAILED(hr))
```

Comment:
DEVICE_STATE_UNPLUGGED adds the duplicates I mentioned.
It's probably old devices or other ports that noone cares about.

I'd just remove them tbh. There's no way of knowing which is which